[WIP] Add DM panel to the left side view.

theViz343 opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR is currently preparation for the addition of a collapsible DM panel in the left side UI view. 


### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [ ] 

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `Private messages sidepanel`
- [x] Fully fixes #1197 
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->
[![asciicast](https://asciinema.org/a/CmHP5b8aFhA6XRLEp0foNJYA9.svg)](https://asciinema.org/a/CmHP5b8aFhA6XRLEp0foNJYA9)

theViz343 committed:
helper/views: Add all_streams unread count to the UnreadCounts dict.

This commit adds an all_streams key to the UnreadCounts dict to store
the unread stream messages count. This is subsequently used by the
StreamPanelButton in the stream panel.

Test case added and tests updated.

theViz343 committed:
buttons/views: Replace stream view with stream panel on left sidebar.

This commit reworks the stream view section in the left sidebar by
replacing it with the new Stream Panel widget. The stream panel consists
of the newly added stream messages button (StreamPanelButton), and the
existing stream/topic view section.

Test added, updated and renamed.

theViz343 committed:
boxes/views: Add context label parameter to PanelSearchBox.

This commit adds a label parameter to the PanelSearchBox to provide
context to the user which searching the streams/topics list.

Tests updated.

theViz343 committed:
model: Fetch recent direct messages from the Zulip API.

This commit adds a list of recent direct messages for the user to the ZT
model (recent_dms). It also adds a _sort_recent_dms function to keep the
list in sorted order according to the max_message_id of the dm.

Fixture added and test updated.

theViz343 committed:
model: Update recent_dms while handling a message_event.

This commit adds an _update_recent_dms method which updates the
recent_dms list when a message event is received. It keeps the list in
sorted order as well.

Tests modified.

theViz343 committed:
buttons/views: Add panel folding for DM and stream panels.

theViz343 committed:
buttons/views:Add DMButton, DMPanel and DMView classes.

theViz343 committed:
views: Rework direct messages UI.

This commit reworks the direct messages button by moving it to a
different section between the menu and the streams panel. It also
adds a direct messages panel which shows all DM recipients.

theViz343 committed:
buttons/views: Rename PMButton to DMPanelButton.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -882,11 +888,9 @@ def topics_view(self, stream_button: Any) -> Any:
         self.view.topic_w = TopicsView(topics_btn_list, self.view, stream_button)
         w = urwid.LineBox(
             self.view.topic_w,
-            title="Topics",
-            title_attr="column_title",
-            tlcorner=COLUMN_TITLE_BAR_LINE,
-            tline=COLUMN_TITLE_BAR_LINE,
-            trcorner=COLUMN_TITLE_BAR_LINE,
```
This looks independent of this particular commit, and maybe belongs in the previous commit?

Also, this change essentially sets all the borders to be empty?

neiljp reviewed (commented):
@theViz343 I'd tested this previously for UX feedback, but not looked at the code in depth, until now :)

The DM Panel logic isn't quite complete functionally from what I remember, so I've focused on seeing if we can get some early commits from this PR which prepare for that work merged sooner, which is the first 3 commits.

Running the PR at that point appears to work broadly fine, with the new separate count and the stream and topic lists working as expected.

As per a comment, I'd drop the shortcut key text until the later DM-related commits, otherwise if we do a partial merge it's confusing - S is still set up to do a contextual change of narrow at that point.

Related to that, functionally there is a nit that we shouldn't be narrowing to all messages from the stream messages button - if anything, we should be narrowing to all non-DMs, but that's not implemented. At this point, you could make the button non-selectable instead. That may be relevant for the later commits, since we've not discussed the distinction between using the hotkeys (P,S) to switch panels vs ones to go to those narrows themselves - like P currently does.

In general the commits I looked at are good, but as per some comments, they might benefit from splitting/reordering. I did manage to trigger an exception when running at the first commit, since the button didn't yet exist.

theViz343 commented:
The recent update to this PR addresses most of the earlier feedback.

The commits are now reordered so that the unread count functionality for the Stream Panel Button is added after the button's introduction.

I've removed the `S` hotkey from the early commits and only introduced it for the Stream Panel Button in the last commit which adds panel folding/unfolding. This might also benefit from some discussion on what to do for the current function of the `S` key (narrow to stream of the current message).

I've also addressed the focus position [issue](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Private.20messages.20sidepanel.20.23T1416/near/1642055). Now, the earlier focused position should be focused on when folding/unfolding panels. 

I've also added the user presence status to the DMButtons. A thing to note is that a DMButton is quite close in UI and functionality, to the UserButton class, with some differences such as the UserInfoView on pressing `i` and narrow to user vs opening compose box for user. This may warrant a parent class (still a child class of TopButton ) from which these two classes inherit. This may simplify the code a bit.

There are still a few tests to be added to cover the new code, but any initial testing is welcome :)
